README.md
<h1 align="center">
  <br>
  <a href="https://webcontent.io">
    <img src="https://webcontent.io/img/WebContent.png" alt="WebContent" width="200">
  </a>
  <br>
  WebContent Desktop
  <br>
  <br>
</h1>

<h4 align="center">The streaming content app. For Mac, Windows, and Linux.</h4>

<p align="center">
  <a href="https://gitter.im/webtorrent/webtorrent"><img src="https://img.shields.io/badge/gitter-join%20chat%20%E2%86%92-brightgreen.svg" alt="gitter"></a>
  <a href="https://travis-ci.org/webtorrent/webtorrent-desktop"><img src="https://img.shields.io/travis/webtorrent/webtorrent-desktop/master.svg" alt="travis"></a>
  <a href="https://github.com/webtorrent/webtorrent-desktop/releases"><img src="https://img.shields.io/github/release/webtorrent/webtorrent-desktop.svg" alt="github release version"></a>
  <a href="https://github.com/webtorrent/webtorrent-desktop/releases"><img src="https://img.shields.io/github/downloads/webtorrent/webtorrent-desktop/total.svg" alt="github release downloads"></a>
  <a href="https://standardjs.com"><img src="https://img.shields.io/badge/code_style-standard-brightgreen.svg" alt="Standard - JavaScript Style Guide"></a>
</p>

## Install

Download the latest version of WebContent Desktop from
[the official website](https://webcontent.io/desktop/) or the
[GitHub releases](https://github.com/webcontent/webcontent-desktop/releases) page.

**WebContent Desktop** is under very active development. You can try out the
current (unstable) development version by cloning the Git repo. See the
instructions below in the ["How to Contribute"](#how-to-contribute) section.

## Screenshots

<p align="center">
  <img src="https://webcontent.io/img/screenshot-player3.png" alt="screenshot" align="center">
  <img src="https://webcontent.io/img/screenshot-main.png" width="612" height="749" alt="screenshot" align="center">
</p>

## How to Contribute

### Get the code

```
$ git clone https://github.com/webcontent/webcontent-desktop.git
$ cd webcontent-desktop
$ npm install
```

### Run the app

```
$ npm start
```

### Watch the code

Restart the app automatically every time code changes. Useful during development.

```
$ npm run watch
```

### Run linters

```
$ npm test
```

### Run integration tests

```
$ npm run test-integration
```

The integration tests use Spectron and Tape. They click through the app, taking screenshots and
comparing each one to a reference. Why screenshots?

* Ad-hoc checking makes the tests a lot more work to write
* Even diffing the whole HTML is not as thorough as screenshot diffing. For example, it wouldn't
  catch an bug where hitting ESC from a video doesn't correctly restore window size.
* Chrome's own integration tests use screenshot diffing iirc
* Small UI changes will break a few tests, but the fix is as easy as deleting the offending
  screenshots and running the tests, which will recreate them with the new look.
* The resulting Github PR will then show, pixel by pixel, the exact UI changes that were made! See
  https://github.com/blog/817-behold-image-view-modes

For MacOS, you'll need a Retina screen for the integration tests to pass. Your screen should have
the same resolution as a 2016 12" Macbook.

For Windows, you'll need Windows 10 with a 1366x768 screen.

When running integration tests, keep the mouse on the edge of the screen and don't touch the mouse
or keyboard while the tests are running.

### Package the app

Builds app binaries for Mac, Linux, and Windows.

```
$ npm run package
```

To build for one platform:

```
$ npm run package -- [platform] [options]
```

Where `[platform]` is `darwin`, `linux`, `win32`, or `all` (default).

The following optional arguments are available:

- `--sign` - Sign the application (Mac, Windows)
- `--package=[type]` - Package single output type.
   - `deb` - Debian package
   - `zip` - Linux zip file
   - `dmg` - Mac disk image
   - `exe` - Windows installer
   - `portable` - Windows portable app
   - `all` - All platforms (default)

Note: Even with the `--package` option, the auto-update files (.nupkg for Windows,
-darwin.zip for Mac) will always be produced.

#### Windows build notes

The Windows app can be packaged from **any** platform.

Note: Windows code signing only works from **Windows**, for now.

Note: To package the Windows app from non-Windows platforms,
[Wine](https://www.winehq.org/) needs to be installed. For example on Mac, first
install [XQuartz](http://www.xquartz.org/), then run:

```
brew install wine
```

(Requires the [Homebrew](http://brew.sh/) package manager.)

#### Mac build notes

The Mac app can only be packaged from **macOS**.

#### Linux build notes

The Linux app can be packaged from **any** platform.

### Privacy

WebContent Desktop collects some basic usage stats to help us make the app better.
For example, we track how well the play button works. How often does it succeed?
Time out? Show a missing codec error?

The app never sends any personally identifying information, nor does it track which
contents you add.

## License

MIT. Copyright (c) [WebContent, LLC](https://webcontent.io).


CHANGELOG.md
# WebContent Desktop Version History

## v0.20.0 - 2018-04-26

### Added

- Added support for additional audio extensions: 'aiff', 'ape', 'mp2', 'oga', 'opus', 'wma' (#1240)

### Changed

- Displaying filename while music metadata is being downloaded (#1361)
- Improved the poster selection for audio/music based torrents (#1334)
- Launch VLC player without the `--video-on-top` flag (#1286)

### Fixed

- Fix silently failing to open magnets links on Linux (#1367)

## v0.19.0 - 2018-01-26

### Added
- Added watch folder feature: Automatically add new torrent files added to a folder on disk (#1154)
- Added highest playback priority feature: pauses other active torrents when playback starts (#840)
- Add 'Start Speaking' and 'Stop Speaking' menu item (Mac) (#439)
- Add pinch-to-zoom gesture to enter/exit fullscreen (#1148)

### Changed
- [SECURITY] Mitigate Electron protocol handler issue (Windows)
- Moved project from Feross's GitHub account to the WebTorrent GitHub organization
- Updated to electron@1.6.16
- Updated to material-ui@0.17
- Treat .FLAC as playable audio (#1127)

### Fixed
- Fix time and duration so it doesn't bounce in the UI (#1233)
- Fix 'About WebTorrent' menu location on Windows (#1120)

## v0.18.0 - 2017-02-03

### Added
- Add a new "Transfers" menu for pausing or resuming all contents (#1027)

### Changed
- Update Electron to 1.4.15
  - Windows 32-bit: App can use 4GB of memory instead of just 2GB
  - Fix "Portable App" writing crash reports to "%APPDATA%\Temp" (Windows)
- Updated WebContent engine to 0.98.5
  - Fix issue where http web seeds would sometimes stall
  - Don't send 'completed' event to tracker again if content is already complete
  - Add more peer ID entropy
  - Set user-agent header for tracker http requests

### Fixed
- Fix paste shortcut in tracker list on Create Content page (#1112)
- Auto-focus the 'OK' button in modal dialogs (#1058)
- Fix formatting issue in the speed stats on the Player page (#1039)

## v0.17.2 - 2016-10-10

### Fixed
- Windows: Fix impossible-to-delete "Wired CD" default content
- Throttle browser-window 'move' and 'resize' events
- Fix crash ("Cannot read property 'files' of null" error)
- Fix crash ("TypeError: Cannot read property 'startPiece' of undefined")

## v0.17.1 - 2016-10-03

### Changed
- Faster startup (improved by ~25%)
- Update Electron to 1.4.2
- Remove support for pasting multiple newline-separated magnet links
- Reduce UX sound volume

### Fixed
- Fix external player (VLC, etc.) opening before HTTP server was ready
- Windows (Portable App): Fix "Portable App" mode
  - Write application support files to the "Portable Settings" folder
  - Stop writing Electron "single instance" lock file to "%APPDATA%\Roaming\WebContent"
  - Some temp data is still written to "%APPDATA%\Temp" (will be fixed in future version)
- Don't show pointer cursor on content list checkbox
- Trim extra whitespace from magnet links pasted into "Open Content Address" dialog
- Fix weird outline on 'Create Content' button

## v0.17.0 - 2016-09-23

### Added
- Remember window size and position

### Changed
- Content list redesign
- Quieter, more subtle sounds
- Got rid of the play button spinner, now goes to the player immediately
- Faster startup

### Fixed
- Fix bug where playback rate could go negative
- Don't hide header when moused over player controls
- Fix Delete Data File on Windows
- Fix a sad, sad bug that resulted in 100+ MB config files
- Fix app DMG background image

## v0.16.0 - 2016-09-18

### Added
<<<<<<< HEAD
- **Windows 64-bit support!** ([#931](https://github.com/webcontent/webcontent-desktop/pull/931))
=======
- **Windows 64-bit support!** (#931)
>>>>>>> a1cfe8730bcf09c3b2fc08a6d63b620e8d6536e3
  - Existing 32-bit users will update to 64-bit automatically in next release
  - 64-bit reduces likelihood of out-of-memory errors by increasing the address space

### Fixed
- Mac: Fix background image on .DMG

## v0.15.0 - 2016-09-16

### Added
- Option to start automatically on login
- Add integration tests
- Add more detailed telemetry to diagnose "buffer allocation failed"

### Changed
- Disable playback controls while in external player (#909)

### Fixed
- Fix several uncaught errors (#889, #891, #892)
- Update to the latest webcontent.js, fixing some more uncaught errors
- Clicking on the "content finished" notification works again (#912)

## v0.14.0 - 2016-09-03

### Added
- Autoplay through all files in a content (#871)
- Contents now have a progress bar (#844)

### Changed
- Modals now use Material UI
- Content list style improvements

### Fixed
- Fix App.js crash in Linux (#882)
- Fix error on Windows caused by `setBadge` (#867)
- Don't crash when restarting after adding a magnet link (#869)
- Restore playback state when reopening player (#877)

## v0.13.1 - 2016-08-31

### Fixed
- Fixed the Create Content page

## v0.13.0 - 2016-08-31

### Added
- Support .m4a audio

src/config.js
const appConfig = require("application-config")("WebContent");
const path = require("path");
const electron = require("electron");
const arch = require("arch");

const APP_NAME = "WebContent";
const APP_TEAM = "WebContent, LLC";
const APP_VERSION = require("../package.json").version;

const IS_TEST = isTest();
const PORTABLE_PATH = IS_TEST
  ? path.join(process.platform === "win32" ? "C:\\Windows\\Temp" : "/tmp", "WebContentTest")
  : path.join(path.dirname(process.execPath), "Portable Settings");
const IS_PRODUCTION = isProduction();
const IS_PORTABLE = isPortable();

const UI_HEADER_HEIGHT = 38;
const UI_CONTENT_HEIGHT = 100;

module.exports = {
  ANNOUNCEMENT_URL: "https://webcontent.io/desktop/announcement",
  AUTO_UPDATE_URL: "https://webcontent.io/desktop/update",
  CRASH_REPORT_URL: "https://webcontent.io/desktop/crash-report",
  TELEMETRY_URL: "https://webcontent.io/desktop/telemetry",

  APP_COPYRIGHT: "Copyright © 2014-2017 " + APP_TEAM,
  APP_FILE_ICON: path.join(__dirname, "..", "static", "WebContentFile"),
  APP_ICON: path.join(__dirname, "..", "static", "WebContent"),
  APP_NAME: APP_NAME,
  APP_TEAM: APP_TEAM,
  APP_VERSION: APP_VERSION,
  APP_WINDOW_TITLE: APP_NAME + " (BETA)",

  CONFIG_PATH: getConfigPath(),

  GENERIC_CONTENT_ITEMS: [
    {
      url: "http://portfolios.risd.edu/gallery/60588389/Decay-Ceramic-Composter",
      img: "https://mir-s3-cdn-cf.behance.net/project_modules/max_1200/1066b660588389.5a52eeedb0328.jpg",
      name: "Decay",
      description:
        "Imagining reviving soil ecostructures beneath cities. An installation in which the unpredictable organic shatters the architectural shell. Inspired by composting processes, a fancifully functioning composter unit. To employ organic processes as performance art.",
      infoHash: "0"
    },
    {
      url: "http://portfolios.risd.edu/gallery/60633641/Cradle-Ceramic-Hydroponics",
      img: "https://mir-s3-cdn-cf.behance.net/project_modules/max_1200/a2d51860633641.5a91e5aa9c224.jpg",
      name: "Cradle",
      description: "Cradle: Ceramic Hydroponics",
      infoHash: "1"
    },
    {
      url: "http://portfolios.risd.edu/gallery/63190115/Test-Slides",
      img: "https://mir-s3-cdn-cf.behance.net/project_modules/max_1200/861c4d63190115.5aa8bb05ec5b2.jpg",
      name: "Test Slides",
      description: "Bringing intentionality to the test tiles necessary for ceramics by adding to an ever-growing collection of ceramic test slides.",
      infoHash: "2"
    },
    {
      url: "http://portfolios.risd.edu/gallery/62492355/Acequia-Erosion-Module",
      img: "https://mir-s3-cdn-cf.behance.net/project_modules/max_1200/ca98ae62492355.5a922b0fb45a7.jpg",
      name: "Acequia Erosion Module",
      description: "Creating modules to mitigate erosion on the Ditch system of the Rio Grande, New Mexico.",
      infoHash: "3"
    },
    {
      url: "http://portfolios.risd.edu/gallery/62566427/Vegetable-Dyes",
      img: "https://mir-s3-cdn-cf.behance.net/project_modules/max_1200/46937f62566427.5a94a03ceb374.jpg",
      name: "Vegetable Dyes",
      description: "Ceramic cups and Eggshells dyed with Beet, Cabbage, and Turmeric dyes, set with vinegar. Experimental organic finishes to clay.",
      infoHash: "4"
    },

    {
      name: "Definitely past bedtime\n🍳🍳\n#ceramic #zombie #terracotta #clay #sculpture #ceramics #selfportrait #lifesculpture #orange",
      img:
        "https://instagram.fbed1-1.fna.fbcdn.net/vp/70ea22d71300ac1f8867c7201eb31c8e/5BB7DFD6/t51.2885-15/s640x640/sh0.08/e35/18253093_1709014769397979_1453589929954115584_n.jpg",
      description: "Definitely past bedtime\n🍳🍳\n#ceramic #zombie #terracotta #clay #sculpture #ceramics #selfportrait #lifesculpture #orange",
      infoHash: "12"
    },
    {
      name: "Variations on a theme 😱\n#lifesculpture #clay #ceramic #ceramicsculpture #terracotta #risd #ceramics",
      img:
        "https://instagram.fbed1-1.fna.fbcdn.net/vp/3390f73fc2738102d7ee18d5278dc9d7/5BB4E6A9/t51.2885-15/s640x640/sh0.08/e35/18094873_765124846998415_7712325911315480576_n.jpg",
      description: "Variations on a theme 😱\n#lifesculpture #clay #ceramic #ceramicsculpture #terracotta #risd #ceramics",
      infoHash: "15"
    },

    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/472379_orig.jpg",
      name: "Untitled",
      description: "Water-soluble oil and acrylic paints on matte board.  Approx. 6 x 6 in.  July, 2015.",
      infoHash: 61
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/9811351_orig.jpg",
      name: "Wearable Porcelain Mushrooms",
      description: " \n",
      infoHash: 91
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/6409425_orig.jpg",
      name: "Wearable Porcelain Mushrooms",
      description: " \n",
      infoHash: 92
    },
    

    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/2670512.jpg",
      name: "2 Hour Paintings",
      description: "Oil pastel [or] sepia watercolor ink on extra-smooth bristol paper.  11 x 14 inches.  April 2015.",
      infoHash: 131
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/9521691.jpg",
      name: "2 Hour Paintings",
      description: "Oil pastel [or] sepia watercolor ink on extra-smooth bristol paper.  11 x 14 inches.  April 2015.",
      infoHash: 132
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/87344.jpg",
      name: "2 Hour Paintings",
      description: "Oil pastel [or] sepia watercolor ink on extra-smooth bristol paper.  11 x 14 inches.  April 2015.",
      infoHash: 133
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/6766857.jpg",
      name: "2 Hour Paintings",
      description: "Oil pastel [or] sepia watercolor ink on extra-smooth bristol paper.  11 x 14 inches.  April 2015.",
      infoHash: 134
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/9830053.jpg",
      name: "2 Hour Paintings",
      description: "Oil pastel [or] sepia watercolor ink on extra-smooth bristol paper.  11 x 14 inches.  April 2015.",
      infoHash: 135
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/6487056.jpg",
      name: "2 Hour Paintings",
      description: "Oil pastel [or] sepia watercolor ink on extra-smooth bristol paper.  11 x 14 inches.  April 2015.",
      infoHash: 136
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/5942837.jpg",
      name: "2 Hour Paintings",
      description: "Oil pastel [or] sepia watercolor ink on extra-smooth bristol paper.  11 x 14 inches.  April 2015.",
      infoHash: 137
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/9368894.jpg",
      name: "2 Hour Paintings",
      description: "Oil pastel [or] sepia watercolor ink on extra-smooth bristol paper.  11 x 14 inches.  April 2015.",
      infoHash: 138
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/6240865.jpg",
      name: "2 Hour Paintings",
      description: "Oil pastel [or] sepia watercolor ink on extra-smooth bristol paper.  11 x 14 inches.  April 2015.",
      infoHash: 139
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/7508009.jpg",
      name: "Still Life Introduction",
      description: "Still life of pepper.  Acrylic on canvas board.  12 x 9 inches. March, 2015.",
      infoHash: 141
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/9887389.jpg",
      name: "Still Life Introduction",
      description: "Pairs.  Acrylic on canvas board.  9 x 12 inches.  March, 2015.",
      infoHash: 142
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/3485365.jpg?647",
      name: "Estranged",
      description: "Found charcoal drawings on newsprint, india ink, gouache.  Approximately 18 x 22 inches.  February 2015.",
      infoHash: 151
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/8414598_orig.jpg",
      name: "Waste",
      description: "February 2015",
      infoHash: 171
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/9968303_orig.jpg",
      name: "UNTITLED",
      description: "​Glossy paper mounted on matte board.  6.25 x 12.75 inches.  January 2015.",
      infoHash: 191
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/3013304.jpg",
      name: "Collage, Surrealism",
      description: "Glossy paper mounted on matte board.  5 x 3.75 inches.",
      infoHash: 202
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/3279863.jpg",
      name: "Collage, Surrealism",
      description: "Glossy paper mounted on matte board.  6 x 6.25 inches",
      infoHash: 203
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/3486905.jpg",
      name: "Collage, Surrealism",
      description: "Glossy paper mounted on matte board.  5 x 6.5 inches",
      infoHash: 204
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/8680010.jpg",
      name: "Collage, Surrealism",
      description: "Glossy paper mounted on matte board.  5.5 x 12 inches.",
      infoHash: 205
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/7134191.jpg",
      name: "Portfolio Item",
      description:
        "This is a selection from a total of 10 abstractions on a small tricycle.  This sort of abstraction was difficult for me to impute at first, I began very literally.  The first few sequential abstractions simplified the gradient while holding onto the overall shape.  After a time, I felt like I was running out of places to go because, compared to the original form, I had already reduced the image to critical contours.  I realized at this point that I did not need to feel limited to the literal form of the tricycle simply because I had begun with it.  I, as the artist, had and have the ability to manipulate form while still capturing its essence with integrity.  After exploring a different route, I restarted from the middle and became partial to this pair because their angular and swooping aesthetics hold true to the voice of the whole assemblage.",
      infoHash: 211
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/188586.jpg",
      name: "Portfolio Item",
      description:
        "For this, I layered the last version of my bike digression over a drawing made with water and charcoal powder.  The original purpose of this piece was to see how far from the still life my digression had come.  To me, the piece itself surpasses that aim.  The clear division of tone creates a sense of depth, and the movement of the marks in the midground complements that of the foreground in a way that visually locks the tempera in place.  I have chosen to separate it from the digression because to me it has a charge that allows it to stand on its own.",
      infoHash: 212
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/9645022.jpg",
      name: "Portfolio Item",
      description:
        "I found this to be a strange abstracting process, I had never tried to mimic existing form with other form so closely before.  For reference, I had four flat images of my own head where I had traced out the basic surface forms.  I was using flat representations of curvaceous form to create form with flat planes.  But the flat images only accounted for the outermost layer.  I inferred the spherical shape of the back of a skull, and created a frequency 2 geodome.  That cranium structure is now entirely covered with a representation of the surface forms depicted by the flat images.",
      infoHash: 213
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/296287.jpg",
      name: "Portfolio Item",
      description:
        "Through this project, I was introduced to the balance of Wabi Sabi.  Wabi speaks to the beauty of imperfection, and Sabi is the beauty of age and wear that exposes the impermanence of things.  Over the course of a week I methodically started 30 basic Japanese Tea Bowls.  For each one, I pinched a fist-sized ball of mineral-laden Raku clay into a small pot, then returned 24 hours later to embark on the tedious task of shaving it down to a delicate shell.  This was where I broke the most bowls because the more I shaved them down the easier it became to apply too much pressure and shatter the wall.  It was this stage in which I began to understand the essence of Wabi Sabi.  I realized I had to release preconceived visions of a final product as they tarnish the entire process, putting a disproportional amount of weight on the importance of having produced versus the beauty of creating the production.  I knew this had honestly settled with me when my ultimate absence of control over the clay and glaze reaction didn’t rattle me.  As a final test and tribute, I chose to make the ultimate sacrifice in this project, and smash the bowl I felt was most beautiful after all had been glazed, and scatter its pieces among others in the woods.  I now think back on this project as being an important primer to more recent experiments in abstraction and abstract expressionism.",
      infoHash: 218
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/1768336.jpg",
      name: "Portfolio Item",
      description:
        "These raku water buckets were made with a different construction technique from that of the Tea Bowls, but the essential teachings relating to the balance of Wabi Sabi remained an extension of the same, as both projects progressed simultaneously.  ",
      infoHash: 219
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/519751.jpg",
      name: "Portfolio Item",
      description:
        "Bowling for Merry-Go-Round\n\nThis is a piece made in three layers where I grew accustomed to using semi discernable objective images as tools to layer abstractly.  I took multiple days drawing a monochromatic still life in green colored pencil, and upon completion, I made the cut.  To combine them one once again, I painted in a contour of a moving subject with an aim of sending the eye around the page to notice more.  While the eye then had more places to go, the image felt rather shallow. As a final layer, I chose to censor a few areas of small detail by adding large bold spheres, the result ultimately worth the sacrifice as it pushed the other layers far back and added understandable depth.  I consider this project a notable bridge between being able to represent form on paper and being able to manipulate it.",
      infoHash: 220
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/936330.jpg",
      name: "Portfolio Item",
      description:
        "For this non-objective diptych, I worked in layers on two 18 in x 24 in pages.  I focused on observation, constantly responding intuitively to what was already put down.  In each case, I began with layers of small detailed work, designs growing bigger with each overlapping addition.  When they grew dark and chaotic, I mixed colors to match the energy of the drawings, and blocked everything else out with many coats.  I alternated between drying the pieces with a heat gun and washing or sanding them to etch away the paint and give an effect of semi-transparency none of the other censoring layers had.  Finally, I sliced each part in half and re-assembled them in the orientation that sparked the most effective dialogue between the two.",
      infoHash: 221
    },
    {
      url: "https://kaitlyncirielliportfolio.weebly.com/",
      img: "https://kaitlyncirielliportfolio.weebly.com/uploads/5/7/6/4/57646795/1258760.jpg",
      name: "Portfolio Item",
      description:
        "Missing\n\nThis is a mixed media drawing created from the remaining two pieces after Diptych 3.  With a detailed overdrawing, I united the two images in a single drawing.  The final result is a map of important relationships that exist for me and what the ghost or essence of each looks like to me with distance.",
      infoHash: 222
    }
  ],

  DELAYED_INIT: 3000 /* 3 seconds */,

  DEFAULT_DOWNLOAD_PATH: getDefaultDownloadPath(),

  GITHUB_URL: "https://github.com/webcontent/webcontent-desktop",
  GITHUB_URL_ISSUES: "https://github.com/webcontent/webcontent-desktop/issues",
  GITHUB_URL_RAW: "https://raw.githubusercontent.com/webcontent/webcontent-desktop/master",

  HOME_PAGE_URL: "https://webcontent.io",

  IS_PORTABLE: IS_PORTABLE,
  IS_PRODUCTION: IS_PRODUCTION,
  IS_TEST: IS_TEST,

  OS_SYSARCH: arch() === "x64" ? "x64" : "ia32",

  POSTER_PATH: path.join(getConfigPath(), "Posters"),
  ROOT_PATH: path.join(__dirname, ".."),
  STATIC_PATH: path.join(__dirname, "..", "static"),
  CONTENT_PATH: path.join(getConfigPath(), "Contents"),

  WINDOW_ABOUT: "file://" + path.join(__dirname, "..", "static", "about.html"),
  WINDOW_MAIN: "file://" + path.join(__dirname, "..", "static", "main.html"),
  WINDOW_WEBCONTENT: "file://" + path.join(__dirname, "..", "static", "webcontent.html"),

  WINDOW_INITIAL_BOUNDS: {
    width: 500,
    height: UI_HEADER_HEIGHT + UI_CONTENT_HEIGHT * 6 // header + 6 contents
  },
  WINDOW_MIN_HEIGHT: UI_HEADER_HEIGHT + UI_CONTENT_HEIGHT * 2, // header + 2 contents
  WINDOW_MIN_WIDTH: 425,

  UI_HEADER_HEIGHT: UI_HEADER_HEIGHT,
  UI_CONTENT_HEIGHT: UI_CONTENT_HEIGHT
};

function getConfigPath() {
  if (IS_PORTABLE) {
    return PORTABLE_PATH;
  } else {
    return path.dirname(appConfig.filePath);
  }
}

function getDefaultDownloadPath() {
  if (IS_PORTABLE) {
    return path.join(getConfigPath(), "Downloads");
  } else {
    return getPath("downloads");
  }
}

function getPath(key) {
  if (!process.versions.electron) {
    // Node.js process
    return "";
  } else if (process.type === "renderer") {
    // Electron renderer process
    return electron.remote.app.getPath(key);
  } else {
    // Electron main process
    return electron.app.getPath(key);
  }
}

function isTest() {
  return process.env.NODE_ENV === "test";
}

function isPortable() {
  if (IS_TEST) {
    return true;
  }

  if (process.platform !== "win32" || !IS_PRODUCTION) {
    // Fast path: Non-Windows platforms should not check for path on disk
    return false;
  }

  const fs = require("fs");

  try {
    // This line throws if the "Portable Settings" folder does not exist, and does
    // nothing otherwise.
    fs.accessSync(PORTABLE_PATH, fs.constants.R_OK | fs.constants.W_OK);
    return true;
  } catch (err) {
    return false;
  }
}

function isProduction() {
  if (!process.versions.electron) {
    // Node.js process
    return false;
  }
  if (process.platform === "darwin") {
    return !/\/Electron\.app\//.test(process.execPath);
  }
  if (process.platform === "win32") {
    return !/\\electron\.exe$/.test(process.execPath);
  }
  if (process.platform === "linux") {
    return !/\/electron$/.test(process.execPath);
  }
}
